File: 189.丢失的第一个正整数/Solution.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def firstMissingPositive(self, A):
        # write your code here
        n = len(A)
        i = 0
        while i < n:
            if A[i] >= 1 and A[i] <= n and A[i] != i+1 and A[A[i]-1] != A[i]:
                A[A[i]-1], A[i] = A[i], A[A[i]-1]
            else:
                i += 1
        n = 1        
        for v in A:
            if v != n:
                break
            n += 1
        return n

File: 189.丢失的第一个正整数/Solution.cpp
class Solution {
public:
    /**    
     * @param A: a vector of integers
     * @return: an integer
     */
    int firstMissingPositive(vector<int> A) {
        // write your code here
        int n = A.size();
        
        for(int i=0; i<n; ){
            if(A[i] == i+1)
                i++;
            else if(A[i]>=1 && A[i]<=n && A[A[i]-1] != A[i])
                swap(A[A[i]-1], A[i]);
            else
                i++;
        }
        
        for(int i=0; i<n; i++)
            if(A[i] != i+1)
                return i+1;
                
        return n+1;
    }
};